service/src/main/kotlin/fi/espoo/evaka/shared/auth/Acl.kt

```diff
@@ -140,6 +140,70 @@ WHERE employee_id = :userId AND decision.id = :decisionId
         }
     )
 
+    fun getRolesForAssistanceNeed(user: AuthenticatedUser, assistanceNeedId: UUID): AclAppliedRoles = AclAppliedRoles(
+        (user.roles - UserRole.ACL_ROLES) + Database(jdbi).read {
+            it.createQuery(
+                // language=SQL
+                """
+SELECT child_id
+FROM assistance_need
+WHERE id = :assistanceNeedId
+                """.trimIndent()
+            ).bind("assistanceNeedId", assistanceNeedId)
+                .mapTo<UUID>()
+                .map { childId -> getRolesForChild(user, childId).roles }
+                .fold(setOf<UserRole>()) { acc, rolesSet -> acc.intersect(rolesSet) }
```

Comment:
Doesn't this always result in an empty set?
We start with an empty set and use intersection -> nothing ever intersects with an empty set so the result will be empty

Reply:
Great catch... We could use `union` here instead.

I don't really like this bit though, since we'd normally expect the set to have exactly one element and the fold feels a bit awkward. Would there be some nice way to express this?

Reply:
I can think of two options:

a)
since the query only returns only one row (or null), you can use something like this:

```
.mapTo<UUID>()
.singleOrNull()
?.let { childId -> getRolesForChild(user, childId).roles }
?: emptySet()
```

b)
Instead of reusing the `getRolesForChild` function, create a reusable database view `child_acl` (employee_id, child_id, role) using a repeatable migration (`R__foo.sql`) and join to it in the query. Then you can have the main query return just `UserRole` instead of having to go through intermediate data

Reply:
Personally I prefer b), because `getRolesForChild` is not really meant to be used by other functions like that. It does two "top-level function" things which are unnecessary when used as a part of another top-level function: it creates a new temporary database connection internally (`Database(jdbi)`) and does the all vs subset check.

Reply:
Thanks for the comments, I'll implement option b.